tests/drivers/spi/qspi/src/main.c

```diff
@@ -0,0 +1,375 @@
+			NULL, (const struct spi_buf_set *)&rx_bufs);
+	zassert_true(ret == 0, "Receive fast read data spi_transceive failure: "
+			"error %d", ret);
+
+	/* check read buffer data whether correct */
+	zassert_true(memcmp(buffer_tx, safbuf2, BUF_SIZE) == 0,
+			"Buffer read data is different to write data");
+
+	spi_cfg_single.operation = SPI_OP_MODE_MASTER | SPI_TRANSFER_MSB
+		| SPI_WORD_SET(8) | SPI_LINES_SINGLE;
+	ret = spi_release(spi_dev, (const struct spi_config *)&spi_cfg_single);
+	zassert_true(ret == 0, "Spi release failure: error %d", ret);
+}
+
+/**
+ * @brief Read data from flash using spi quad mode
+ * @details
+ * - check and make sure spi quad mode is enabled
+ * - read data using spi quad mode
+ * - check read buffer data whether correct
+ */
+void test_spi_quad_read(void)
+{
+	int ret;
+	uint8_t spi_status2;
+
+	/* read register2 to judge whether quad mode is enabled */
+	safbuf[0] = SPI_READ_STATUS2;
+	txb.buf = &safbuf;
+	txb.len = 1U;
+
+	tx_bufs.buffers = (const struct spi_buf *)&txb;
+	tx_bufs.count = 1U;
+
+	memset(safbuf2, 0, 1U);
+	rxb.buf = &safbuf2;
+	rxb.len = 1U;
+
+	rx_bufs.buffers = (const struct spi_buf *)&rxb;
+	rx_bufs.count = 1U;
+
+	ret = spi_transceive(spi_dev,
+			(const struct spi_config *)&spi_cfg_single,
+			(const struct spi_buf_set *)&tx_bufs,
+			(const struct spi_buf_set *)&rx_bufs);
+	zassert_true(ret == 0, "Read register2 status spi_transceive failure: "
+			"error %d", ret);
+
+	spi_status2 = safbuf2[0];
+	/* set register2 QE=1 to enable quad mode */
+	if ((spi_status2 & SPI_STATUS2_QE) == 0U) {
+		safbuf[0] = SPI_WRITE_ENABLE_VS;
+
+		txb.buf = &safbuf;
+		txb.len = 1U;
+
+		tx_bufs.buffers = (const struct spi_buf *)&txb;
+		tx_bufs.count = 1U;
+
+		rx_bufs.buffers = NULL;
+		rx_bufs.count = 0U;
+
+		ret = spi_transceive(spi_dev,
+				(const struct spi_config *)&spi_cfg_single,
+				(const struct spi_buf_set *)&tx_bufs,
+				(const struct spi_buf_set *)&rx_bufs);
+		zassert_true(ret == 0,
+				"Send write enable volatile spi_transceive failure: "
+				"error %d", ret);
+
+		safbuf[0] = SPI_WRITE_STATUS2;
+		safbuf[1] = spi_status2 | SPI_STATUS2_QE;
+
+		txb.buf = &safbuf;
+		txb.len = 2U;
+
+		tx_bufs.buffers = (const struct spi_buf *)&txb;
+		tx_bufs.count = 1U;
+
+		rx_bufs.buffers = NULL;
+		rx_bufs.count = 0U;
+
+		ret = spi_transceive(spi_dev,
+				(const struct spi_config *)&spi_cfg_single,
+				(const struct spi_buf_set *)&tx_bufs,
+				(const struct spi_buf_set *)&rx_bufs);
+		zassert_true(ret == 0,
+				"Write spi status2 QE=1 spi_transceive failure: "
+				"error %d", ret);
+	}
+
+	/* configure spi quad mode */
+	spi_cfg_quad.frequency = TEST_FREQ_HZ;
+	spi_cfg_quad.operation = SPI_OP_MODE_MASTER | SPI_TRANSFER_MSB
+		| SPI_WORD_SET(8) | SPI_LINES_QUAD;
+	spi_cfg_quad.slave = 0;
+	spi_cfg_quad.cs = NULL;
+
+	/* read data using spi quad mode */
+	safbuf[0] = SPI_QUAD_FAST_READ_DATA;
+	safbuf[1] = 0x10U;
+	safbuf[2] = 0x0U;
+	safbuf[3] = 0x0U;
+	safbuf[4] = 0x0U;
```

Comment:
Same comment as for dual.  Pins driven by controller and target during dummy clocks.

Reply:
done